### ISSUE TITLE
Automatic uploading and publishing

### ISSUE THREAD

**Issue by porst17** [IMAGINARY/imaginary-web#61]
I am currently revising my development and deployment infrastructure to be as automatic as possible. Nevertheless, I would have to upload all the resulting binaries manually to the platform. In a continuous integration world (which is my probably unreachable ultimate goal), this would be possible automatic.

I currently use a similar approach for keeping the embedded Cinderella applets up to date: When I push to github, a webhook is called on imaginary.org to `git pull` the changes immediately and notifies me about the changes (+succes or error) via email. But I don't know if and how this approach can work for regular file uploads (which also should include a version number).

**Comment by elondaits** [comment]
This issue is related to this (now) closed request: https://github.com/IMAGINARY/imaginary-web/issues/60

**Comment by elondaits** [comment]
My thoughts:

As per the request in https://github.com/IMAGINARY/imaginary-web/issues/60 , out of the box Drupal doesn't support an "upload from an URL" which wouldn't be an upload but rather mean that the web server downloads the file from the URL. There might be some module that does this, or perhaps one can be programmed (not trivial) but it has some inherent difficulties:

1. Security issues. It's not optimal from a security standpoint for the web server to be able to access arbitrary URLs. Even if only privileged users can do it the security has to be low enough to allow it. Even then it can be considered... it's just something to think about.

2. While the URL is begin fetched by the server the page has to remain open so that when it finishes the user can submit the node. The user has to wait anyway for the whole thing to be fetched while some AJAX mechanism gives feedback on the download. If there's any problem there has to be a clear error reporting mechanism for the user to be able to retry. Again... none of this is impossible, but it's not an easy issue. Even if some 3rd party module that does this exists it's hard enough to do it right that I'd be hard press to trust its quality.

But then... as per this issue it'd be apparently more optimal to do a background fetch process than an interactive one, which still has issue 1 but it's much easier regarding issue 2 (no progress report, no interactive feedback, just a log message in the end in case of success or N failed retries).

Now, Is it necessary for the files to be uploaded to the Imaginary site for the users to download them from it? Can't they instead be linked from github?  I know Github has an API so perhaps whenever a new binary is uploaded a listing in Imaginary can be refreshed with the new release and link... this would be totally continuous, with a minimal delay and no user interaction required.

**Comment by porst17** [comment]
1. I don't see why fetching URLs is less secure than downloading the file from the URL to my local system and uploading it again to the platform besides the effort that is required by the user. I admit, that it is not a good idea to allow every peace of code to fetch data from anywhere. But I don't see an additional security risk during the upload process.

2. Uploading 100MB from my local system takes half an hour, sometimes longer. Fetching it from another webserver (like GitHub) will probably take less than a minute. So waiting time is relative. I did not say, that this is an easy feature request. Nevertheless, I found [this](https://www.drupal.org/project/filefield_sources) within 20 seconds. Looks promising to me.

Actually, we are discussing things that should have been discussed in #60. Maybe #60 should be reopened.

Regarding the actual question of this issue: Automatic fetching of files should be a background process with minimal API. For me, it is not absolutely required to actually upload something to the IMAGINARY platform. It would be sufficient to provide links to the actual files on GitHub (or whereever) which are in turn displayed in the files section, if this complies with GitHubs terms of use. But so far it looks like the IMAGINARY platform wants to be as self contained as possible. It is more a conceptional question and I haven't been involved in related discussion yet. Maybe @christophknoth can provide some insight.

Personally, I would be happy with external links for files but someone has to keep them up to date and remove dead links.

**Comment by elondaits** [comment]
The reason I think it's better to discuss #60 and this together is that if an alternate solution like using the github API or a background process are chosen then there's no need for #60, so both issues have to be considered together.

1. A standard secure server doesn't connect to any other outside servers and in fact might not even have the modules to do so installed if they're not needed (e.g. curl). Of course allowing users to upload files is a HUGE security concern (Drupal has a lot of features to alleviate the problem), but it's a completely different one... so having two security concerns is worse than having one. Like I said, it's not something impossible, but the security implications have to be considered. 

2. Like I said, I'm not surprised a Drupal module exists for this... but it's a module with 80 open issues that hasn't been updated since 2013. Drupal modules are of hugely variable quality and even if they're good they will always open up the door for incompatibilities. It's certainly a possibility to use the module, but if there's an alternate / better solution like an automated system that doesn't require interactivity (which is the hardest part here) I'd certainly lean towards that one.

As for the possibility of external links I propose, if possible, using the github API so it's fully automatic. At intervals the server would acess the github API, fetch the list of releases (and the last one) and use them to build the links. Whenever an update is done on github the site would be updated as soon as the cache gets invalidated (any logical interval... four hours, an hour, 15 minutes, etc.)

**Comment by elondaits** [comment]
I want to be a bit clearer about the security implication of point 1 because I might sound a bit vague:

Whenever you allow a user to place files in the server (e.g. through an upload) there's a huge number of potential security concerns. Some of them are:
- The user could upload a PHP file and then open it, which basically let's him take over the server.
- ... or upload a .js or .swf file, and bypass the "same domain" restrictions which are basic for web security.
- The user could upload a file named something like "../../settings.php" which lets him place files outside the "safe and non executable" uploads directory. He could also attempt to overwrite a file.
- The file could have a name that does something bad when expanded by a UNIX script using wildcards like *.

... there are so many concerns it's impossible to trust one single programmer to consider them all. For uploads Drupal should be safe based on the number of versions, years of development, security updates and eyes on it.

... An alternate method of placing files in the server, like through an URL which the server fetches, would have all the same concerns but would not necessarily take advantage of Drupal's security measures, since it's a different thing altogether. Drupal DOES have features where a file is fetched from a server (module and language pack updates), but they're all meant for admins, and are considered insecure anyway.

Besides that having the server fetch URLs on the user's request has a couple more security concerns I can think of which are not similar to the file upload case:
- The server might be able to access URLs which are not publicly available

**Comment by porst17** [comment]
Well, even if there we can solve the automation process this issue is about, there are still a lot of use cased for #60.

I understand the security concerns, but except for the last point (access non-public URLs) this is still very similar to regular file uploads. I agree that using a module that hasn't been tested that well might pose a problem. But the [FieldField Sources](https://www.drupal.org/project/filefield_sources) module is just an improvement of the [FieldField](https://www.drupal.org/project/filefield) core module.
 * in both cases, the [same guy](https://www.drupal.org/u/quicksketch) is the main author and maintainer
 * both modules have a lot of open issues (80 vs. 71)
 * both modules have no recent commits (452 days ago vs. 251 days ago)

I still see your point, but I am asking myself if it would be possible to restrict usage of this module to admins. It would be a huge time saver for me and probably also @andreasdanielmatt (and of course everybody who does a lot of updates in the files section) and most of these people are not regular users but trusted co-workers. Another nice feature of this module is the ability to select files which have previously been uploaded. Currently, I have to reupload the file and it get's a suffix assigned (at least I think so, haven't tested it recently). 

Regarding the automation process with github: How would I implement this on the platform? GitHub webhooks are pretty straight forward to use, but I don't know how to create links in the files section (neither in the UI nor using any API).

**Comment by elondaits** [comment]
Yes, certainly restricting the functionality to certain trusted people is better.

As for automating the whole thing (which I assume is ultimately preferable since it's closer to a continuous integration model and requires no extra intervention whatsoever): 

GitHub lets you download a .zip or .tar.gz from any release. What I assume (haven't checked) is that I can get the URL to those .zip files through GitHub's API... so what I'd need to do is a module that allows me to associate the Imaginary "Program" node with a GitHub repository's URL (or more than one). This module would generate either the link to download the .zip file from the last release (I could even throw in a regexp to match stable releases vs dev ones) or a list of all the releases.

... I know it's not "as simple as that" as I see in Imaginary's Surfer page that there are many files involved, which probably come from different sources. 

Do you think such a solution would work and remove the need for manually uploading things?

**Comment by porst17** [comment]
1. Do you think to use this module for restricted users only? Not all software on the platform is going to be deployed automatically and even for Surfer I will need quite some time to set everything up. Hence, the possibility to upload from a URL is still of great use for the content maintainers.

2. It sounds like an appropriate solution for automatic deployment to retrieve the URL from Github directly in the short run. But we should not only target Github. Github is for source code and the "release" feature does nothing more than creating a version tag for a specific commit. Although it is possible to add binaries, this is a manual process. Platforms like Bintray are probably more suitable for our needs since they actually target hosting binaries (or at least packaged stuff) and continuous integration (for example using a REST API). But this has to be evaluated. First tests can be run with GitHub. We can later add more platforms for download link generation.
